File: problems/1_two-sum/main.cpp
#include <vector>
#include <unordered_map>
#include <iostream>
#include <sstream>
using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> mp;

        for(int i=0; i< nums.size(); i++){
            int need = target - nums[i];
            if(mp.find(need) != mp.end()){
                return {mp[need], i};
            }
            mp[nums[i]] = i;
        }
        return {};
    }
};

string vec_to_string(const vector<int>& nums) {
    stringstream ss;
    ss << "[";
    for (size_t i = 0; i < nums.size(); ++i) {
        if (i > 0) {
            ss << ", ";
        }
        ss << nums[i];
    }
    ss << "]";
    return ss.str();
}

int main() {
    Solution sol;
    vector<vector<int>> samples = {
        {2, 7, 11, 15},
        {3, 2, 4},
        {3, 3}
    };
    vector<int> targets = {9, 6, 6};
    vector<vector<int>> expected = {
        {0, 1},
        {1, 2},
        {0, 1}
    };

    for (size_t i = 0; i < samples.size(); ++i) {
        vector<int> nums = samples[i];
        vector<int> result = sol.twoSum(nums, targets[i]);
        cout << "Sample " << (i + 1) << ": nums = " << vec_to_string(samples[i])
             << ", target = " << targets[i] << "\n";
        cout << "Placeholder result: " << vec_to_string(result) << "\n";
        cout << "Expected output: " << vec_to_string(expected[i]) << "\n\n";
    }

    return 0;
}

File: problems/1_two-sum/main.rs
use std::collections::HashMap;
struct Solution;

impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        // TODO: implement this method
        let mut map = HashMap::new();
        for (i, &num) in nums.iter().enumerate() {
            let need = target - num;
            if let Some(&idx) = map.get(&need) {
                return vec![idx as i32, i as i32];
            }
            map.insert(num, i);
        }
        Vec::new()
    }
}

fn main() {
    let samples = vec![(vec![2, 7, 11, 15], 9), (vec![3, 2, 4], 6), (vec![3, 3], 6)];
    let expected = vec![vec![0, 1], vec![1, 2], vec![0, 1]];

    for (i, (nums, target)) in samples.iter().enumerate() {
        let result = Solution::two_sum(nums.clone(), *target);
        println!("Sample {}: nums = {:?}, target = {}", i + 1, nums, target);
        println!("Placeholder result: {:?}", result);
        println!("Expected output: {:?}\n", expected[i]);
    }
}
